src/main/java/org/wickedsource/docxstamper/processor/repeat/RepeatDocPartProcessor.java

```diff
@@ -1,233 +1,186 @@
+					os.forEach(o -> recursivelyReplaceImages(o, replacements));
+					os.forEach(c -> setParentIfPossible(c, gcp));
+					changes.addAll(os);
+				}
+			}
+
+			List<Object> gcpContent = gcp.getContent();
+			int index = gcpContent.indexOf(repeatElements.get(0));
+			gcpContent.addAll(index, changes);
+			gcpContent.removeAll(repeatElements);
+		}
+	}
+
+	private Deque<WordprocessingMLPackage> stampSubDocuments(List<Object> subContexts, WordprocessingMLPackage subTemplate) {
+		Deque<WordprocessingMLPackage> subDocuments = new ArrayDeque<>();
+		for (Object subContext : subContexts) {
+			WordprocessingMLPackage templateCopy = outputWord(os -> copy(subTemplate, os));
+			WordprocessingMLPackage subDocument = outputWord(os -> stamp(subContext,
+																		 templateCopy,
+																		 os
+			));
+			subDocuments.add(subDocument);
+		}
+		return subDocuments;
+	}
+
+	private static List<Object> documentAsInsertableElements(WordprocessingMLPackage subDocument, boolean oddNumberOfBreaks, SectPr previousSectionBreak) {
+		List<Object> inserts = new ArrayList<>(DocumentUtil.allElements(subDocument));
+		// make sure we replicate the previous section break before each repeated doc part
+		if (oddNumberOfBreaks && previousSectionBreak != null) {
+			if (DocumentUtil.lastElement(subDocument) instanceof P p) {
+				SectionUtil.applySectionBreakToParagraph(previousSectionBreak, p);
+			} else {
+				// when the last element to be repeated is not a paragraph, we need to add a new
+				// one right after to carry the section break to have a valid xml
+				P p = objectFactory.createP();
+				SectionUtil.applySectionBreakToParagraph(previousSectionBreak, p);
+				inserts.add(p);
+			}
+		}
+		return inserts;
+	}
+
+	private static void recursivelyReplaceImages(Object o, Map<R, R> replacements) {
+		Queue<Object> q = new ArrayDeque<>();
+		q.add(o);
+		while (!q.isEmpty()) {
+			Object current = q.remove();
+			if (replacements.containsKey(current)
+					&& current instanceof Child child
+					&& child.getParent() instanceof ContentAccessor parent) {
+				List<Object> parentContent = parent.getContent();
+				parentContent.add(parentContent.indexOf(current), replacements.get(current));
+				parentContent.remove(current);
+			} else if (current instanceof ContentAccessor ca) {
+				q.addAll(ca.getContent());
+			}
+		}
+	}
+
+	private static void setParentIfPossible(Object object, ContentAccessor parent) {
+		if (object instanceof Child child)
+			child.setParent(parent);
+	}
+
+	private WordprocessingMLPackage outputWord(Consumer<OutputStream> outputter) {
+		try (
+				PipedOutputStream os = new PipedOutputStream();
+				PipedInputStream is = new PipedInputStream(os)
+		) {
+			Thread thread = THREAD_FACTORY.newThread(() -> outputter.accept(os));
+			thread.start();
+			WordprocessingMLPackage wordprocessingMLPackage = WordprocessingMLPackage.load(is);
+			thread.join();
```

Comment:
This seems to make multithreading useless, since you're joining the thread you created in the same method, it will run synchronously. Is there a reason I'm missing ?

Reply:
You don't have to make an intermediate bytearray in memory, so it should run in about half the time and use less RAM. But I agree I need to measure that theoretical gain before merging.